### PR TITLE
Normz Coding Challenge

### PR DESCRIPTION
**Pull request by normancapule**

### CODE REVIEW

File: app/controllers/todos_controller.rb [hunk 1]
```diff
@@ -4,12 +4,13 @@ def new
...
 
   def create
-    run Todo::Create do
-      return redirect_to @model.list, notice: "To-do was successfully created!"
+    todo = run Todo::Create
+    if todo.errors.any?
+      flash.now[:alert] = todo.errors.full_messages.join(",")
+      render :new
+    else
+      return redirect_to todo.model.list, notice: "To-do was successfully created!"
```
Comment (iamjasonaevans) [inline app/controllers/todos_controller.rb:12]: Good stuff. FYI The preferred pattern for this trailblazer is 

```
def create
    run Comment::Create do |op|
         flash[:notice] = "Created comment for \"#{op.thing.name}\"" return redirect_to thing_path(op.thing)
    end

    @thing = Thing.find(params[:thing_id]) # UI-specific logic!
    render :new 
end
```

File: app/concepts/shared/common.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    ##### otherwise, create a new one
+    if params[:current_user].to_s.blank?
+      if params[:current_user_id].to_s.blank?
+        User::Create.(user: {}).model
+      else
+        User.find(params[:current_user_id])
+      end
+    else
+      params[:current_user]
+    end
```
Comment (iamjasonaevans) [inline app/concepts/shared/common.rb:16]: I think a guard clause could work in this situation a little better.

```
return params[:current_user] unless params[:current_user].to_s.blank?

if params[:current_user_id].to_s.blank?
  User::Create.(user: {}).model
else
  User.find(params[:current_user_id])
end
```

File: app/concepts/todo/create.rb [hunk 3]
```diff
@@ -1,50 +1,40 @@
...
+  model Todo, :create
 
-    # find or create the parent list for the todo
-    if params[:todo_list_id].blank?
-      todo_list = TodoList.find_or_create_by(name: "Default To-do List")
+  contract do
+    property :title
+    property :description
+    property :todo_list_id
+    validates :title, presence: true
```
Comment (iamjasonaevans) [inline app/concepts/todo/create.rb:11]: 👍

**Review by iamjasonaevans — COMMENTED** [review]
Hi Norman,

I had a quick look at your PR. Overall a good attempt. Have you used Trailblazer before?

File: app/concepts/user/create.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    property :username
+    property :fullname
+  end
+
+  def process(params)
+    validate(params[:user]) do
+      contract.fullname ||= "Guest"
+      contract.save
+    end
+    invalid!
```
Comment (iamjasonaevans) [inline app/concepts/user/create.rb:15]: We can probably remove this line.

File: app/concepts/todo_list/create.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def process(params)
+    validate(params[:todo_list]) do
+      if contract.user_id.blank?
+        contract.user_id = current_user(params).id
+      end
+      contract.save
+    end
+    invalid!
```
Comment (iamjasonaevans) [inline app/concepts/todo_list/create.rb:19]: We can probably remove this line.

**Comment by normancapule** [comment]
Hi @iamjasonaevans. This was my first time using trailblazer.

File: app/concepts/shared/common.rb (hunk 2 above)
Comment (normancapule) [reply app/concepts/shared/common.rb:16]: Yeah. Makes sense. Didn't occur to me to use Policy::Guard here.

**Pull request closed without merging**